consensus/polybft/governance_manager.go

```diff
@@ -0,0 +1,411 @@
+)
+
+const (
+	oldRewardLookbackSize = uint64(2) // number of blocks to calculate commit epoch info from the previous epoch
+	newRewardLookbackSize = uint64(1)
+)
+
+var errUnknownGovernanceEvent = errors.New("unknown event from governance")
+
+// isRewardDistributionBlock indicates if reward distribution transaction
+// should happen in given block
+// if governance fork is enabled, reward distribution is only present on the first block of epoch
+// and if we are not at the start of chain
+// if governance fork is not enabled, reward distribution is only present at the epoch ending block
+func isRewardDistributionBlock(isFirstBlockOfEpoch, isEndOfEpoch bool,
+	pendingBlockNumber uint64) bool {
+	if forkmanager.GetInstance().IsForkEnabled(chain.Governance, pendingBlockNumber) {
+		return isFirstBlockOfEpoch && pendingBlockNumber > 1
+	}
+
+	return isEndOfEpoch
+}
+
+// getLookbackSizeForRewardDistribution returns lookback size for reward distribution
+// based on if governance fork is enabled or not
+func getLookbackSizeForRewardDistribution(blockNumber uint64) uint64 {
+	if forkmanager.GetInstance().IsForkEnabled(chain.Governance, blockNumber) {
+		return newRewardLookbackSize
+	}
+
+	return oldRewardLookbackSize
+}
+
+// GovernanceManager interface provides functions for handling governance events
+// and updating client configuration based on executed governance proposals
+type GovernanceManager interface {
+	PostBlock(req *PostBlockRequest) error
+	PostEpoch(req *PostEpochRequest) error
+	GetClientConfig() (*PolyBFTConfig, error)
+}
+
+var _ GovernanceManager = (*dummyGovernanceManager)(nil)
+
+// dummyStakeManager is a dummy implementation of GovernanceManager interface
+// used only for unit testing
+type dummyGovernanceManager struct{}
+
+func (d *dummyGovernanceManager) PostBlock(req *PostBlockRequest) error { return nil }
+func (d *dummyGovernanceManager) PostEpoch(req *PostEpochRequest) error { return nil }
+func (d *dummyGovernanceManager) GetClientConfig() (*PolyBFTConfig, error) {
+	return nil, nil
+}
+
+var _ GovernanceManager = (*governanceManager)(nil)
+
+// governanceManager is a struct that saves governance events
+// and updates the client configuration based on executed governance proposals
+type governanceManager struct {
+	logger       hclog.Logger
+	state        *State
+	eventsGetter *eventsGetter[contractsapi.EventAbi]
+}
+
+// newGovernanceManager is a constructor function for governance manager
+func newGovernanceManager(genesisConfig *PolyBFTConfig,
+	logger hclog.Logger,
+	state *State,
+	blockhain blockchainBackend) (*governanceManager, error) {
+	eventsGetter := &eventsGetter[contractsapi.EventAbi]{
+		blockchain: blockhain,
+		isValidLogFn: func(l *types.Log) bool {
+			return l.Address == contracts.NetworkParamsContract
+		},
+		parseEventFn: parseGovernanceEvent,
+	}
+
+	config, err := state.GovernanceStore.getClientConfig()
+	if config == nil || errors.Is(err, errClientConfigNotFound) {
```

Comment:
What if `getClientConfig` returns an error? Should we propagate it from `newGovernanceManager` (in else if branch)?

Reply:
Fix: https://github.com/0xPolygon/polygon-edge/pull/1749/commits/6bcffbbd344e5db00b863d22c1872ace2a23a9f6